cpp/src/column/column_view.cpp

```diff
@@ -80,21 +80,32 @@ size_type column_view_base::null_count(size_type begin, size_type end) const
 
 // simple prime number multiplication algorithm.
 // Adapted from http://myeyesareblind.com/2017/02/06/Combine-hash-values/#apachecommons
-constexpr void combine_hash(size_t& h1, size_t h2) { h1 = h1 * 127 + h2; }
+constexpr void combine_hash(std::size_t& h1, std::size_t h2) { h1 = h1 * 127 + h2; }
+
+struct shallow_hash_impl {
+  std::size_t operator()(column_view const& input, bool is_parent_empty = false)
+  {
+    std::size_t hash = 0;
+    combine_hash(hash, std::hash<data_type>{}(input.type()));
+    combine_hash(hash, std::hash<size_type>{}(input.size()));
+    if (not(input.is_empty() or is_parent_empty)) {
+      combine_hash(hash, std::hash<void const*>{}(input.head()));
+      combine_hash(hash, std::hash<void const*>{}(input.null_mask()));
+      combine_hash(hash, std::hash<size_type>{}(input.offset()));
+    }
+    hash = std::accumulate(
+      input.child_begin(),
+      input.child_end(),
+      hash,
+      [&input, is_parent_empty](std::size_t hash, auto const& child) {
+        combine_hash(hash, shallow_hash_impl{}(child, input.is_empty() or is_parent_empty));
+        return hash;
+      });
+    return hash;
+  }
+};
```

Comment:
I don't understand why the `is_parent_empty` logic is required. How a column is hashed seems like it should be independent of the state of it's parent. 

I think we can simplify this a bit though. 

```suggestion
constexpr std::size_t combine_hash(std::size_t h1, std::size_t h2) {return h1 * 127 + h2; }

template <typename... Ts>
constexpr auto hash(Ts&&... ts){
   return (std::hash<Ts>{ts} ^ ... );
}

std::size_t shallow_hash(column_view const& c)
  {
    auto const init = c.is_empty() ? hash(c.type(), c.size()) : hash(c.type(), c.size(), c.head(), c.null_mask(), c.offset());
    
    return std::accumulate(
      input.child_begin(),
      input.child_end(),
      init,
      [&c](std::size_t hash, auto const& child) {
        return combine_hash(hash, shallow_hash(child));
      });
  }
```

Reply:
Also, I'm not sure I understand why the `offset` is only included if the column isn't empty. That means two empty columns with different offsets would be treated as shallow equal. That doesn't seem right.

Reply:
could we use different hash function than XOR?
we could use boost hash_combine or other fast candidates from 
http://myeyesareblind.com/2017/02/06/Combine-hash-values/#apachecommons

`combine_hash` explained here at https://github.com/rapidsai/cudf/pull/9185#discussion_r704098492